### DISCUSSION TITLE
Script to create a custom package for a given project from all public packages

### DISCUSSION THREAD

**Discussion by satya4satyanm** [npm/feedback#1016] [General]
Lately I was thinking is there a way to create a custom package of the code snippets picked from multiple other packages for a project need.

Problem statement: Due to transitive dependency, we face unsupported version issue or the deprecated warnings when something is changed. And project migration is very difficult when there is a major version released for a parent library like react or angular.

Benefits of having custom package for a project:
1. It will pull in the necessary code snippet from other package to use a specific feature and store in custom package
2. Transitive dependencies can be avoided when a package is distributed
3. Build process won't need to download the packages
4. Developers will have more control to extend a snippet by over-riding the methods
5. May be with a minor transpile into JavaScript, generalising a component can be achieved by which the same component can be used by multiple libraries like react. angular, vue, etc because everything is end of the day JavaScript.

We still need the public packages from which the code snippets will be pulled in through a script line like `npm create custom-library`. But those packages just have to create their custom library of code snippets instead of transitive dependencies.

Today we see each package in npmJs is dependent on hundreds of other packages. Out of which one might be used heavily and others are used for a specific purpose. So to support this feature, in package.json there should be a section where we will define the package list from which we need snippets only.

For example,
For lodash package - let us say I am using a specific method (_.sortedUniq(array)) 
Why lodash should be a dependency for my project. I just need that method signature in my custom package.

lodash still can be an item in my package.json file in a different section so that we know a particular method is extracted from a particular package.

Hope this is clear.

For now this is just an idea and need more discussion.

**Comment by ljharb** [comment]
A tool to automate copy-pasting seems like a great way to violate licenses, and lose the ability to be notified about security issues.

If you want a subset of a package, ask them to extract it to a separate package, or do so yourself if the license permits it.
